### PR TITLE
[WIP] Magic Ball Task

### PR DESCRIPTION
**Pull request by VasiliZas**
Magic 8 ball — шар, внешне напоминающий бильярдный шар № 8, но большего размера — обычно диаметром 10-11 см, внутри которого есть ёмкость с тёмной жидкостью, например, чернилами. В жидкости плавает фигура с 20 поверхностями — икосаэдр, на каждой из которых нанесено по одному ответу. держа шар в руках (окошком вниз) пользователь задаёт ему вопрос, встряхивает, переворачивает его и читает появившийся ответ.

Ответы нанесены в формате «да», «нет», «абсолютно точно», «плохие шансы», «вопрос не ясен», и т. д. Всего 20 вариантов ответов.

 Написать алгоритм, который в случайном порядке выбирает фразу из коллекции.

### CODE REVIEW

File: homework/Vasili_Zasinets/src/main/java/homework/ball/Magic.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        ball.put(15, "Don’t count on it");
+        ball.put(16, "My reply is no");
+        ball.put(17, "My sources say no");
+        ball.put(18, "Outlook not so good");
+        ball.put(19, "Very doubtful");
+        ball.put(20, "Concentrate and ask again");
+    }
+
+    int number() {
+        return random.nextInt(21);
```
Comment (PavelBoPavel) [inline homework/Vasili_Zasinets/src/main/java/homework/ball/Magic.java:45]: +

**Review by PavelBoPavel — APPROVED** [review]

**Review by PavelBoPavel — DISMISSED** [review]
lgfm

**Review by trump-ny — APPROVED** [review]

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+public class Magic {
+    Logger log = LoggerFactory.getLogger("Magic Ball ");
+    Map<Integer, String> ball = new HashMap<>();
+    Random random = new Random();
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:13]: privat static and maybe final

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+public class Magic {
+    Logger log = LoggerFactory.getLogger("Magic Ball ");
+    Map<Integer, String> ball = new HashMap<>();
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: privat

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+public class Magic {
+    Logger log = LoggerFactory.getLogger("Magic Ball ");
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:11]: Why is there a space at the end?

**Review by KaydunovDenis — CHANGES_REQUESTED** [review]

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 3 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: Why?

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 2 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:13]: Why?

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    int number() {
+        return random.nextInt(20);
+    }
+
+    void prediction() {
+        log.info("Prediction fo you: {}", ball.get(number()));
+    }
+
```
[the remark this answers is not in the document]
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:52]: Why?

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 3 above)
Comment (KaydunovDenis) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: Best practice explicitly specify the access modifier for methods and varible of class

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 4 above)
Comment (KaydunovDenis) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:11]: Empty in end is "костыль" change your logger propeties

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 2 above)
Comment (KaydunovDenis) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:13]: Best practice explicitly specify the access modifier for methods and varible of class.
Veriable "random" is constant of class.

Constant class Java
A constant is a variable whose value cannot change once it has been assigned. Java doesn't have built-in support for constants, but the variable modifiers static and final can be used to effectively create one. Constants can make your program more easily read and understood by others.

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (KaydunovDenis) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:45]: Best practice explicitly specify the access modifier for methods and varible of class.

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public String toString() {
+        return "Magic{" +
+                "ball=" + ball +
+                '}';
+    }
+
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:52]: According to last page java cod convetion you need delete empty line.

File: homework/Vasili_Zasinets/src/test/java/com/tutrit/java/quickstart/homework/ball/MagicTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    Logger log = LoggerFactory.getLogger("Ball ");
+
+
+    @Before
+    public void setUp() {
+        this.magic = new Magic();
+    }
+
+    @Test
+    public void addPrediction() {
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/test/java/com/tutrit/java/quickstart/homework/ball/MagicTest.java:19]: These test is true always.
I thing it need rewrite.
You can use ball.size()

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        ball.put(17, "Outlook not so good");
+        ball.put(18, "Very doubtful");
+        ball.put(19, "Concentrate and ask again");
+    }
+
+    int number() {
+        return random.nextInt(20);
+    }
+
+    void prediction() {
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:42]: According Java Convention name of methods should be the verb.
prediction - > predict
Use hot keys: Shift+F6

**Review by KaydunovDenis — CHANGES_REQUESTED** [review]

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:45]: fixed

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 4 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:11]: fixed

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 3 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: fixed

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 2 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:13]: fixed

File: homework/Vasili_Zasinets/src/main/java/homework/ball/Magic.java (hunk 1 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/homework/ball/Magic.java]: fixed

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 6 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:52]: fixed

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 8 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:42]: fixed

File: homework/Vasili_Zasinets/src/test/java/com/tutrit/java/quickstart/homework/ball/MagicTest.java (hunk 7 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/test/java/com/tutrit/java/quickstart/homework/ball/MagicTest.java:19]: fixed

**Review by KaydunovDenis — APPROVED** [review]

**Review by lamaconda — DISMISSED** [review]
lgfm

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+public class Magic {
+    private final Random random = new Random();
+    private final Map<Integer, String> ball = new HashMap<>();
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: I think what you need rename it variable
ball->predicts

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    Logger log = LoggerFactory.getLogger("Magic Ball");
+
+    public void addPredict() {
```
Comment (KaydunovDenis) [inline homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:15]: I think what you need rename it methods
addPredict->
initialPredicts
or
addPredicts

..something like this

**Review by KaydunovDenis — CHANGES_REQUESTED** [review]

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 10 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:15]: Thanks. I think this name is good. I will not change it.

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 9 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:12]: Thanks. I think this name is good. I will not change it.

File: homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
+package com.tutrit.java.quickstart.homework.ball;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+public class Magic {
```
[the remark this answers is not in the document]
Comment (VasiliZas) [reply homework/Vasili_Zasinets/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:10]: Thanks. I think this name is good. I will not change it.

**Review by KaydunovDenis — DISMISSED** [review]

**Review by PavelBoPavel — APPROVED** [review]

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        ball.put(12, "Better not tell you now");
+        ball.put(13, "Cannot predict now");
+        ball.put(14, "Don’t count on it");
+        ball.put(15, "My reply is no");
+        ball.put(16, "My sources say no");
+        ball.put(17, "Outlook not so good");
+        ball.put(18, "Very doubtful");
+        ball.put(19, "Concentrate and ask again");
+    }
+
```
Comment (lamaconda) [inline homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:37]: cool!

**Review by lamaconda — DISMISSED** [review]
lgfm

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        ball.put(10, "Reply hazy,try again");
+        ball.put(11, "Ask again later");
+        ball.put(12, "Better not tell you now");
+        ball.put(13, "Cannot predict now");
+        ball.put(14, "Don’t count on it");
+        ball.put(15, "My reply is no");
+        ball.put(16, "My sources say no");
+        ball.put(17, "Outlook not so good");
+        ball.put(18, "Very doubtful");
+        ball.put(19, "Concentrate and ask again");
```
Comment (max-java) [inline homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:35]: It's a good example to be put in initialization block, like we did in enum lecture.

Please, try.

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public int number() {
+        return random.nextInt(20);
```
Comment (max-java) [inline homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:39]: use map size as upper bound. 
`random.nextInt(ball.size());`

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 14 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java:39]: ok

**Review by KaydunovDenis — DISMISSED** [review]

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java (hunk 13 above)
Comment (VasiliZas) [reply homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Magic.java]: ok

**Review by KaydunovDenis — APPROVED** [review]

**Comment by VasiliZas** [comment]
MagicBall1
MagicBall2

**Review by KaydunovDenis — APPROVED** [review]

File: homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Ball.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
+package com.tutrit.java.quickstart.homework.ball;
+
+public enum Ball {
```
Comment (max-java) [inline homework/Vasili_Zasinets/MagicBallTask/src/main/java/com/tutrit/java/quickstart/homework/ball/Ball.java:3]: no need to create enum for that. Prediction collection or map was a good decision.
let's revise this on next techTalk using SOLID?

**Review by max-java — CHANGES_REQUESTED** [review]
ptal

**Pull request closed without merging**